### PR TITLE
StringConversion.cpp: rewrite the StringFromUTF8 function. Closes #1193

### LINKED ISSUE #1193: [CoP] Build Release doesn't starts on Arch Linux after mimalloc update
**Describe the bug**
On latest dev branch I cannot run CoP. Engine crashes with segmentation fault.

**To Reproduce**
1. Update dev branch
2. Configure Release build
3. Compile Release
4. Run
5. See logs

**Expected behavior**
The game start successful

**Screenshots**
-

**log**
<details><summary>openxray_ruut.log</summary>

```
! Couldn't create surface from image: Parameter 'src' is invalid
OpenXRay Release build 8724, Dec 26 2022 (Linux 64-bit, shared)
Custom build from commit[dee5323de20e6815250cee7da3187e6406b5891a] branch[dev]
 
command line -nointro -fsltx /run/media/ruut/99c9b836-a7b8-403c-9ef8-5bfc1e1ec831/games/stalker/CoP/fsgame.ltx 

* CPU features: RDTSC, MMX, SSE, AVX
* CPU threads: 8

Initializing File System...
$fs_root$ = /run/media/ruut/99c9b836-a7b8-403c-9ef8-5bfc1e1ec831/games/stalker/CoP
using fs-ltx /run/media/ruut/99c9b836-a7b8-403c-9ef8-5bfc1e1ec831/games/stalker/CoP/fsgame.ltx
FS: 39682 files cached 28 archives, 20Kb memory used.
Init FileSystem 0.398881 sec
-----loading \run\media\ruut\99c9b836-a7b8-403c-9ef8-5bfc1e1ec831\games\stalker\CoP\gamedata\configs\system.ltx
-----loading \run\media\ruut\99c9b836-a7b8-403c-9ef8-5bfc1e1ec831\games\stalker\CoP\gamedata\configs\system.ltx
! Failed to determine on which monitor the game is launched.
! SDL: Invalid window
Starting INPUT device...
Loading module: xrRender_GL
Available render modes[1]:
renderer_rgl
Executing config-script "user.ltx"...
[\run\media\ruut\99c9b836-a7b8-403c-9ef8-5bfc1e1ec831\games\stalker\CoP\appdata\user.ltx] successfully loaded.
Selected renderer: renderer_rgl
Loading module: xrGame
Initializing Engine...
SOUND: OpenAL: enumerate devices...
SOUND: OpenAL: EnumerationExtension Present
SOUND: OpenAL: system default sound device name is OpenAL Soft
SOUND: OpenAL: All available devices:
1. OpenAL Soft, Spec Version 1.1 (default) eax[0] efx[yes] xram[no]
Executing config-script "\run\media\ruut\99c9b836-a7b8-403c-9ef8-5bfc1e1ec831\games\stalker\CoP\gamedata\configs\default_controls.ltx"...
 
FATAL ERROR
 
[error] Expression    : <no expression>
[error] Function      : handler_base
[error] File          : /home/ruut/Documents/qemu_shared/xray-16/src/xrCore/xrDebug.cpp
[error] Line          : 866
[error] Description   : segmentation fault
 

stack trace:

xrDebug::GatherInfo(char*, unsigned long, ErrorLocation const&, char const*, char const*, char const*, char const*)
xrDebug::Fail(bool&, ErrorLocation const&, char const*, char const*, char const*, char const*)
/home/ruut/Documents/qemu_shared/xray-16/bin/x64/Release/xrCore.so(+0x3f8f7) [0x7f62e93458f7]
/usr/lib/libc.so.6(+0x38a00) [0x7f62e8b69a00]
operator delete(void*)
StringFromUTF8[abi:cxx11](char const*, std::locale const&)
KbdKeyToButtonName(int, std::__cxx11::basic_string<char, std::char_traits<char>, xalloc<char> >&)
CInput::GetKeyName(int, char*, int)
/home/ruut/Documents/qemu_shared/xray-16/bin/x64/Release/xrEngine.so(+0x8f22f) [0x7f62e955e22f]
CCC_Bind::Execute(char const*)
CConsole::ExecuteCommand(char const*, bool)
CCC_LoadCFG::Execute(char const*)
CConsole::ExecuteCommand(char const*, bool)
CCC_DefControls::Execute(char const*)
CConsole::ExecuteCommand(char const*, bool)
execUserScript()
Startup()
RunApplication()
./bin/x64/Release/xr_3da(+0x16ad) [0x557172ead6ad]
./bin/x64/Release/xr_3da(+0x13ef) [0x557172ead3ef]
```

</details>

**Desktop (please complete the following information):**
 - OS: Arch linux 
 - OpenXRay build version commit: dee5323de20e6815250cee7da3187e6406b5891a

**Additional context**
```
$ uname -a
Linux ruut 6.0.12-zen1-1-zen #1 ZEN SMP PREEMPT_DYNAMIC Thu, 08 Dec 2022 11:03:40 +0000 x86_64 GNU/Linux
```

```
$ yay -Qsq mimalloc
local/mimalloc 2.0.7-1
```

### PR DESCRIPTION
**Pull request by ChugunovRoman**

### CODE REVIEW

**Comment by Xottab-DUTY** [comment]
Is it possible to get rid of std::locale usage at all?

**Comment by ChugunovRoman** [comment]
> Is it possible to get rid of std::locale usage at all?

I think no. I could't find any information about an alternative option for converting locale characters

File: src/xrCore/Text/StringConversion.cpp [hunk 1]
```diff
@@ -158,16 +159,23 @@ u16 mbhMulti2Wide(xr_wide_char* WideStr, xr_wide_char* WidePos, u16 WideStrSize,
...
-    xr_string result(wstr.size(), '\0');
-    std::use_facet<std::ctype<wchar_t>>(locale).narrow(wstr.data(), wstr.data() + wstr.size(), '?', &result[0]);
+    int len = 0;
+    auto size = utf8len(in);
+    xr_string result(size, '\0');
+    utf8_int32_t cp;
+    in = utf8codepoint(in, &cp);
+    while('\0' != cp)
+    {
+        result[len] = std::use_facet<std::ctype<wchar_t>>(locale).narrow((wchar_t)cp, '?');
```
Comment (dreamer-dead) [inline src/xrCore/Text/StringConversion.cpp:171]: Shouldn't the `len` variable incremented then?

**Comment by Xottab-DUTY** [comment]
This doesn't release us from the use of std::locale and deprecated functionality.

I researched some more about this. It looks like, that [ICU](https://github.com/unicode-org/icu) is the best choice.

**Pull request closed without merging**